physical/azure/azure_ha.go

```diff
@@ -0,0 +1,422 @@
+	defer ticker.Stop()
+
+	for {
+		// writeLock intentionally not place within case<-ticker to avoid the
+		// initial retry interval wait
+		acquired, err := l.writeLock()
+		if err != nil {
+			return false, errwrap.Wrapf("attempt lock: {{err}}", err)
+		}
+		if acquired {
+			return true, nil
+		}
+
+		select {
+		case <-ticker.C:
+		case <-stopCh:
+			return false, nil
+		}
+	}
+}
+
+// renewLock renews the given lock until the channel is closed.
+func (l *Lock) renewLock() {
+	ticker := time.NewTicker(l.renewInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			l.writeLock()
+		case <-l.stopCh:
+			return
+		}
+	}
+}
+
+// watchLock checks whether the lock has changed in the table and closes the
+// leader channel accordingly. If an error occurs during the check, watchLock
+// will retry the operation and then close the leader channel if it can't
+// succeed after retries.
+func (l *Lock) watchLock() {
+	retries := 0
+	ticker := time.NewTicker(l.watchRetryInterval)
+
+OUTER:
+	for {
+		// Check if the channel is already closed
+		select {
+		case <-l.stopCh:
+			break OUTER
+		default:
+		}
+
+		// Check if we've exceeded retries
+		if retries >= l.watchRetryMax-1 {
+			break OUTER
+		}
+
+		// Wait for the timer
+		select {
+		case <-ticker.C:
+		case <-l.stopCh:
+			break OUTER
+		}
+
+		// Attempt to read the key
+		r, err := l.get(context.Background())
+		if err != nil {
+			retries++
+			continue
+		}
+
+		// Verify the identity is the same
+		if r == nil || r.Identity != l.identity {
+			break OUTER
+		}
+	}
+
+	l.stopLock.Lock()
+	defer l.stopLock.Unlock()
+	if !l.stopped {
+		l.stopped = true
+		close(l.stopCh)
+	}
+}
+
+// writeLock writes the given lock using the following algorithm:
+//
+// - lock does not exist
+//   - write the lock
+// - lock exists
+//   - if key is empty or identity is the same or timestamp exceeds TTL
+//     - update the lock to self
+func (l *Lock) writeLock() (bool, error) {
+	l.backend.permitPool.Acquire()
+	defer l.backend.permitPool.Release()
+
+	// Create a transaction to read and the update (maybe)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// The operation may be retried, so we need to stop it if we lose leadership.
+	go func() {
+		select {
+		case <-l.stopCh:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+
+	// Build up the access conditions
+	var conds azblob.BlobAccessConditions
+
+	// Read the record
+	r, err := l.get(ctx)
+	if err != nil {
+		return false, errwrap.Wrapf("write lock: {{err}}", err)
+	}
+	if r != nil {
+		// If the key is empty or the identity is ours or the ttl expired, we can
+		// write. Otherwise, return now because we cannot.
+		if r.Key != "" && r.Identity != l.identity && time.Now().UTC().Sub(r.Timestamp) < l.ttl {
+			return false, nil
+		}
+		conds.IfMatch = r.etag // CAS mechanism
+	} else {
+		// IfNoneMatch: "*" = perform the operation iff the resource does not exist
```

Comment:
Just spotted a small typo.

```suggestion
		// IfNoneMatch: "*" = perform the operation if the resource does not exist
```

Reply:
this is short for "if and only if"